tests/wpt/web-platform-tests/performance-timeline/po-observe.any.js

```diff
@@ -33,9 +40,11 @@
   }, "Filter unsupported entryType entryType names within the entryTypes sequence");
 
   async_test(function (t) {
-  var observer = new PerformanceObserver(
+    var observer = new PerformanceObserver(
       t.step_func(function (entryList, obs) {
-        assert_equals(observer, obs, "observer is second parameter");
+        assert_equals( observer.observe({entryTypes: ["mark"]}), 
+          obs.observe({entryTypes: ["measure"]}),
+          "observer is second parameter. replace the old observer's options with the new one.");    
```

Comment:
[observer.observe()](https://github.com/servo/servo/blob/master/components/script/dom/webidls/PerformanceObserver.webidl#L20) has no return value, so we are essentially comparing `undefined` to `undefined` here. That's why you see the test passing.

Reply:
Is it possible to check the "options" mentioned in W3C have been replaced?

Reply:
Yes, that is exactly what the existing test is doing.

The [options](https://w3c.github.io/performance-timeline/#dom-performanceobserverinit) parameter is the one containing the [entryTypes](https://w3c.github.io/performance-timeline/#dom-performanceobserverinit-entrytypes) array where the observer defines for which type of performance entries it wants its callback to be called.

The test is setting two different `options` **for the same [observer](https://github.com/servo/servo/blob/master/tests/wpt/web-platform-tests/performance-timeline/po-observe.any.js#L36)**. The [first time](https://github.com/servo/servo/blob/master/tests/wpt/web-platform-tests/performance-timeline/po-observe.any.js#L46) it is asking the observer to observe performance entries of type `mark`. The [second time](https://github.com/servo/servo/blob/master/tests/wpt/web-platform-tests/performance-timeline/po-observe.any.js#L47) it is asking it to observe entries of type `measure`. Given step 3 of the [PerfomanceObserver.observe()](https://w3c.github.io/performance-timeline/#dom-performanceobserver-observe()) spec, we should replace the first `options` parameter with the second one. So when [self.performance.mark("mark1")](https://github.com/servo/servo/blob/master/tests/wpt/web-platform-tests/performance-timeline/po-observe.any.js#L48) is executed, the observer callback must **not** be called. But when [self.performance.measure("measure1")](https://github.com/servo/servo/blob/master/tests/wpt/web-platform-tests/performance-timeline/po-observe.any.js#L49) is executed, the observer callback must be called. And it should be called with an `entryList` containing only the performance entry of type `measure`, but not the one of type `mark`.

The options replacement code is [here](https://github.com/servo/servo/blob/master/components/script/dom/performance.rs#L114). What's probably wrong is the way we check if one observer is the same as another one.

Reply:
@ferjm  thanks for your kind explanation. 

I have some questions:
- name `mark1` is not defined at all?
- Is it required to test `self.performance.mark("mark1")` observer callback being called? If not, why do we need this line?

Reply:
If comment one of it, the original version test is failed.
```
observer.observe({entryTypes: ["mark"]});
// observer.observe({entryTypes: ["measure"]});
```
or 
```
// observer.observe({entryTypes: ["mark"]});
observer.observe({entryTypes: ["measure"]});
```

Reply:
If comment one of it, the original version test is also failed.
```
observer.observe({entryTypes: ["mark"]});
// observer.observe({entryTypes: ["measure"]});
```
or
```
// observer.observe({entryTypes: ["mark"]});
observer.observe({entryTypes: ["measure"]});
```

Reply:
> name mark1 is not defined at all?

`mark1` is a string [here](https://github.com/servo/servo/blob/master/tests/wpt/web-platform-tests/performance-timeline/po-observe.any.js#L48). It is the name of the performance entry of type `mark` that the test is adding. Note that even though we are recording this performance entry, the observer will ignore it, because it is only interested in entries of type `measure` at that point. That's why it is not part of the list [here](https://github.com/servo/servo/blob/master/tests/wpt/web-platform-tests/performance-timeline/po-observe.any.js#L39)

> Is it required to test self.performance.mark("mark1") observer callback being called? If not, why do we need this line?

The observer callback must **not** be called for `self.performance.mark("mark1")`. We need this line to make sure that the list of entries that we check after receiving the callback for `self.performance.measure("measure1")` does not contain the entry of type `mark` corresponding to `self.performance.mark("mark1")`. Which is the expected behavior.

Reply:
> If comment one of it, the original version test is failed.

Note that what you see is probably the test behaving as expected. I guess what you are seeing is something like:

```ssh
```

What you are running here are [Web Platform Tests](https://github.com/w3c/web-platform-tests), which is a test suite shared by all major browser vendors. What you do when running this command:

Reply:
> If comment one of it, the original version test is failed.

I guess what you are seeing is something like:

```ssh
❯ ./mach test-wpt tests/wpt/web-platform-tests/performance-timeline
Running 13 tests in web-platform-tests

Ran 13 tests finished in 13.0 seconds.
  • 13 ran as expected. 0 tests skipped.
```

Note that what it says is that the tests run as expected, not that they failed or succeeded.

What you are running here are [Web Platform Tests](https://github.com/w3c/web-platform-tests), which is a test suite shared by all major browser vendors. What you do when running this command: `/mach test-wpt tests/wpt/web-platform-tests/performance-timeline`, is to run the performance-timeline test suite and compare the results with the results that we expect. By default we expect that the tests pass, but if we expect any failures, we have to write this expectation in a `.ini` file specific for the failing test.

In this case, the test expectation is a [failure](https://github.com/servo/servo/pull/18283/files#diff-f260931d5371d798a71d39395c0befe9R4). The goal of this PR should be to turn this failure into a success, by changing **the implementation** of PerformanceObserver. So one thing that you'll need to do at some point is to remove the `.ini` file where we say that we expect this test to fail, which is `tests/wpt/metadata/performance-timeline/po-observe.any.js.ini`. Removing this file will let you test your changes in the implementation easily.

Reply:
Sorry for not clear.
I removed `.ini`.
and change test to 
```
//observer.observe({entryTypes: ["mark"]});
observer.observe({entryTypes: ["measure"]});
```

get 
```
tigercosmos@ubuntu:~/Desktop/servo$ ./mach test-wpt tests/wpt/web-platform-tests/performance-timeline
Running 1 tests in web-platform-tests

  ▶ Unexpected subtest result in /performance-timeline/po-observe.any.worker.html:
  │ FAIL [expected PASS] replace observer if already present
  │   → assert_equals: expected object "[object PerformanceObserver]" but got object "[object PerformanceObserver]"
  │ 
  │ observer<@http://web-platform.test:8000/performance-timeline/po-observe.any.js:45:9
  │ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1406:20
  └ Test.prototype.step_func/<@http://web-platform.test:8000/resources/testharness.js:1430:20

  ▶ Unexpected subtest result in /performance-timeline/po-observe.any.html:
  │ FAIL [expected PASS] replace observer if already present
  │   → assert_equals: expected object "[object PerformanceObserver]" but got object "[object PerformanceObserver]"
  │ 
  │ observer<@http://web-platform.test:8000/performance-timeline/po-observe.any.js:45:9
  │ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1406:20
  └ Test.prototype.step_func/<@http://web-platform.test:8000/resources/testharness.js:1430:20

Ran 13 tests finished in 17.0 seconds.
  • 11 ran as expected. 0 tests skipped.
  • 2 tests had unexpected subtest results
```

I think it is related to implement of `PerformanceObserver.observe()`, because when I do this, it is no matter about step3.

Reply:
Ok, I see what you mean. It is complaining about the PerformanceObserver given to the callback not being the expected one. Which makes sense because of https://github.com/servo/servo/blob/master/components/script/dom/performance.rs#L185

I just checked the [latest version](https://github.com/w3c/web-platform-tests/blob/master/performance-timeline/po-observe.any.js) of this test in the upstream repo and it seems that [this assertion](https://github.com/servo/servo/blob/master/tests/wpt/web-platform-tests/performance-timeline/po-observe.any.js#L38) has been removed. So I guess we can safely remove it for Servo as well until we sync the tests  upstream.

Reply:
Shall we end #18285 and #18372? Or what's next?

Reply:
#18285 is still relevant as the issue has not been fixed yet. The fact that the assertion mentioned above is not valid does not mean that the rest of the test is wrong and we are still failing the test, so we should fix that.

Reply:
Sorry, I am confused.

Should we keep figuring out why makes the assertion mentioned above failed, which other browser can pass,  by checking our `PerformanceObserver.observe()` implement.
Or because of  the latest version of this test in the upstream just remove it,  so we sync the tests upstream and then check the test to see what happen?

Reply:
> Should we keep figuring out why makes the assertion mentioned above failed, which other browser can pass, by checking our PerformanceObserver.observe() implement?

No. Removing the assertion from [the test](https://github.com/servo/servo/blob/master/tests/wpt/web-platform-tests/performance-timeline/po-observe.any.js#L38) is fine. We will eventually sync our copy of WPTs upstream, but that's outside of the scope of this issue.

In summary, what needs to be done here is:

1. Fix the fact that we don't replace an already registered observer options when we should. This likely happens because the comparison between observers [here](https://github.com/servo/servo/blob/master/components/script/dom/performance.rs#L120) is wrong. Try to compare observers directly, not references. Something like `*o.observer == *observer`.
2. Remove the [assertion](https://github.com/servo/servo/blob/master/tests/wpt/web-platform-tests/performance-timeline/po-observe.any.js#L38) from the test, as we know it has been removed upstream and it is no longer valid.
3. [Update the test expectations](https://github.com/servo/servo/tree/master/tests/wpt#updating-test-expectations), because after 1. and 2. the test should pass.
4. [Update the WPT manifest](https://github.com/servo/servo/tree/master/tests/wpt#updating-the-wpt-manifest), because you changed the test in 2.

Reply:
I try to compare reference `*o.observer == *observer`, but the `assert_equals(observer, obs)` in `po-observe.any.js` still failed.
Seems not so easy.

Reply:
That's step 2. of the [list](https://github.com/servo/servo/pull/18372#discussion_r137236933) above :) You need to remove that assertion.

Reply:
Although we will remove that line, I thought it should pass, because you tested that line should be OK in Firefox and Chrome before?

Reply:
We can remove that line because it's already gone [upstream](https://github.com/w3c/web-platform-tests/blob/master/performance-timeline/po-observe.any.js#L58) and yes, the test is passing in Firefox and Chrome without that line.

Reply:
OK, I got the points XD

Reply:
Good!

It seems that you still need to update the test expectations (point 3 of the list).

Also, please rebase on top of master and squash the commits into a single one.

Reply:
@ferjm  In don't understand this, " In this case, remove the references to tests whose expectation is now PASS, and remove .ini files that no longer contain any expectations." from the [document](https://github.com/servo/servo/tree/master/tests/wpt#updating-test-expectations).

How to determine "no longer contain any expectations"? I thought if the `.ini` only have "PASS" would be the condition?

Reply:
> It seems that you still need to update the test expectations (point 3 of the list).

Nevermind. It seems that you already removed the test expectations in 5eabe5a03b2671802ed274b85d9d523ecd1b913c. Sorry for the confusion.